src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs

```diff
@@ -0,0 +1,573 @@
+        /// namespce to be changed. If this reference is the right side of a qualified name, the new node returned would
+        /// be the entire qualified name. Depends on whether <paramref name="newNamespaceParts"/> is provided, the name 
+        /// in the new node might be qualified with this new namespace instead.
+        /// </summary>
+        /// <param name="reference">A reference to a type declared inside the namespce to be changed, which is calculated 
+        /// based on results from `SymbolFinder.FindReferencesAsync`.</param>
+        /// <param name="newNamespaceParts">If specified, the namespace of original reference will be replaced with given 
+        /// namespace in the replacement node.</param>
+        /// <param name="old">The node to be replaced. This might be an ancestor of original </param>
+        /// <param name="new">The replacement node.</param>
+        public abstract bool TryGetReplacementReferenceSyntax(
+            SyntaxNode reference, ImmutableArray<string> newNamespaceParts, ISyntaxFactsService syntaxFacts, out SyntaxNode old, out SyntaxNode @new);
+
+        protected abstract TCompilationUnitSyntax ChangeNamespaceDeclaration(
+            TCompilationUnitSyntax root, ImmutableArray<string> declaredNamespaceParts, ImmutableArray<string> targetNamespaceParts);
+
+        protected abstract SyntaxList<TMemberDeclarationSyntax> GetMemberDeclarationsInContainer(SyntaxNode compilationUnitOrNamespaceDecl);
+
+        protected static SyntaxAnnotation WarningAnnotation { get; }
+            = CodeActions.WarningAnnotation.Create(
+                FeaturesResources.Warning_colon_changing_namespace_may_produce_invalid_code_and_change_code_meaning);
+
+        /// <summary>
+        /// This code action tries to change the name of the namespace declaration to 
+        /// match the folder hierarchy of the document. The new namespace is constructed 
+        /// by concatenating the default namespace of the project and all the folders in 
+        /// the file path up to the project root.
+        /// 
+        /// For example, if he default namespace is `A.B.C`, file path is 
+        /// "[project root dir]\D\E\F\Class1.cs" and declared namespace in the file is
+        /// `Foo.Bar.Baz`, then this action will change the namespace declaration
+        /// to `A.B.C.D.E.F`. 
+        /// 
+        /// Note that it also handles the case where the target namespace or declared namespace 
+        /// is global namespace, i.e. default namespace is "" and the file is located at project 
+        /// root directory, and no namespace declaration in the document, respectively.
+        /// </summary>
+        public async Task<Solution> ChangeNamespaceAsync(Solution solution, ImmutableArray<DocumentId> documentIds, string declaredNamespace, string targetNamespace, CancellationToken cancellationToken)
```

Comment:
need to document what DocumentIds is for.

Reply:
I will be making changes to this after this PR is merged, so leaving it as is.